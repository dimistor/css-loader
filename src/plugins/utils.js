import postcss from 'postcss';

const createICSSImports = (imports) =>
  Object.keys(imports).map((path) => {
    const aliases = imports[path];
    const declarations = Object.keys(aliases).map((key) =>
      postcss.decl({
        prop: key,
        value: aliases[key],
        raws: { before: '\n  ' },
      })
    );

    return postcss
      .rule({
        selector: `:import(${path})`,
        raws: { after: '\n' },
      })
      .append(declarations);
  });

export default { createICSSImports };
