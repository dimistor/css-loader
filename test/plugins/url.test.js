import fs from 'fs';
import path from 'path';

import plugin from '../../src/plugins/url';
import { runPostcss } from '../helpers/utils';

describe('URL postcss plugin', () => {
  it('basic', async () => {
    const result = await runPostcss(
      fs.readFileSync(path.join(__dirname, '../fixtures/url/url.css')),
      [plugin]
    );

    expect(result.css).toMatchSnapshot('css');
  });
});
