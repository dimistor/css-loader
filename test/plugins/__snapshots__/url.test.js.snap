// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`URL postcss plugin basic: css 1`] = `
":import('img.png') {
  __url__0__: default;
}

:import('img.png#hash') {
  __url__1__: default;
}

:import('~package/img.png') {
  __url__2__: default;
}

:import('other-img.png') {
  __url__3__: default;
}

:import('img img.png') {
  __url__4__: default;
}

:import('font.woff') {
  __url__5__: default;
}

:import('font.woff2') {
  __url__6__: default;
}

:import('font.eot') {
  __url__7__: default;
}

:import('~package/font.ttf') {
  __url__8__: default;
}

:import('font.eot?#iefix') {
  __url__9__: default;
}

:import('font.svg#svgFontName') {
  __url__10__: default;
}

:import('img2x.png') {
  __url__11__: default;
}

:import('./url-imported.css') {
  __url__12__: default;
}

.class {
  background: url(__url__0__);
}

.class-hash {
  background: url(__url__1__);
}

.class {
  background: url(
    __url__0__
  );
}

.class {
  background: green url(__url__0__);
}

.class {
  background: url(__url__0__) xyz;
}

.class {
  background: green url(__url__0__) xyz;
}

.class {
  background: url( __url__0__ );
}

.class {
  background: url(__url__0__);
}

.class {
  background: url(__url__0__);
}

.class {
  background: url(
    __url__0__
  );
}

.class {
  background: url( __url__0__ );
}

.class {
  background: url(__url__2__) url(__url__3__);
}

.class {
  background: green url(__url__2__) url(__url__3__);
}

.class {
  background: url(__url__2__) url(__url__3__) xyz;
}

.class {
  background: green url( __url__2__ ) url(__url__3__) xyz;
}

.class {
  background: url( __url__4__ );
}

.class {
  background: url( __url__4__ );
}

.class {
  background: url(/img.png);
}

.class {
  background: url(data:image/png;base64,AAA) url(http://example.com/image.jpg) url(//example.com/image.png);
}

.class {
  background: url(data:image/png;base64,AAA);
}

.class {
  background: url(\\"data:image/svg+xml;charset=utf-8,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 42 26' fill='%23007aff'><rect width='4' height='4'/><rect x='8' y='1' width='34' height='2'/><rect y='11' width='4' height='4'/><rect x='8' y='12' width='34' height='2'/><rect y='22' width='4' height='4'/><rect x='8' y='23' width='34' height='2'/></svg>\\");
}

.class {
  background: url(\\"data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns%3D%27http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%27%20viewBox%3D%270%200%2042%2026%27%20fill%3D%27%2523007aff%27%3E%3Crect%20width%3D%274%27%20height%3D%274%27%2F%3E%3Crect%20x%3D%278%27%20y%3D%271%27%20width%3D%2734%27%20height%3D%272%27%2F%3E%3Crect%20y%3D%2711%27%20width%3D%274%27%20height%3D%274%27%2F%3E%3Crect%20x%3D%278%27%20y%3D%2712%27%20width%3D%2734%27%20height%3D%272%27%2F%3E%3Crect%20y%3D%2722%27%20width%3D%274%27%20height%3D%274%27%2F%3E%3Crect%20x%3D%278%27%20y%3D%2723%27%20width%3D%2734%27%20height%3D%272%27%2F%3E%3C%2Fsvg%3E\\")
}

.class {
  filter: url('data:image/svg+xml;charset=utf-8,<svg xmlns=\\"http://www.w3.org/2000/svg\\"><filter id=\\"filter\\"><feGaussianBlur in=\\"SourceAlpha\\" stdDeviation=\\"0\\" /><feOffset dx=\\"1\\" dy=\\"2\\" result=\\"offsetblur\\" /><feFlood flood-color=\\"rgba(255,255,255,1)\\" /><feComposite in2=\\"offsetblur\\" operator=\\"in\\" /><feMerge><feMergeNode /><feMergeNode in=\\"SourceGraphic\\" /></feMerge></filter></svg>#filter');
}

.class {
  filter: url('data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns%3D%5C%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%5C%22%3E%3Cfilter%20id%3D%5C%22filter%5C%22%3E%3CfeGaussianBlur%20in%3D%5C%22SourceAlpha%5C%22%20stdDeviation%3D%5C%220%5C%22%20%2F%3E%3CfeOffset%20dx%3D%5C%221%5C%22%20dy%3D%5C%222%5C%22%20result%3D%5C%22offsetblur%5C%22%20%2F%3E%3CfeFlood%20flood-color%3D%5C%22rgba(255%2C255%2C255%2C1)%5C%22%20%2F%3E%3CfeComposite%20in2%3D%5C%22offsetblur%5C%22%20operator%3D%5C%22in%5C%22%20%2F%3E%3CfeMerge%3E%3CfeMergeNode%20%2F%3E%3CfeMergeNode%20in%3D%5C%22SourceGraphic%5C%22%20%2F%3E%3C%2FfeMerge%3E%3C%2Ffilter%3E%3C%2Fsvg%3E%23filter');
}

.class {
  filter: url(#highlight);
}

.class {
  filter: url('#line-marker');
}

.class {
  filter: url(\\"#line-marker\\");
}

@font-face {
  src: url(__url__5__) format('woff'),
    url(__url__6__) format('woff2')
    url(__url__7__) format('eot'),
    url(__url__8__) format('truetype');
}

@font-face {
  src: url(__url__9__) format('embedded-opentype');
}

@font-face {
  src: url(__url__10__) format('svg')
}

@media (min-width: 500px) {
  .class {
    background: url(__url__0__);
  }
}

.class {
  content: \\"do not use url(path)\\";
}

.class {
  content: 'do not \\"use\\" url(path)';
}

@keyframes anim {
  background: url(__url__0__);
}

.class {
  background: -webkit-image-set(url(__url__0__) 1x, url(__url__11__) 2x);
}

.class {
  background: url();
}

.class {
  background: url('');
}

.class {
  background: url('   ');
}

.class {
  background: url(
    ' '
  );
}

.class {
  background: url(https://raw.githubusercontent.com/webpack/media/master/logo/icon.png);
}

.class {
  background: url(//raw.githubusercontent.com/webpack/media/master/logo/icon.png);
}

.foo {
  color: red;
}

.imported-css {
  background: url(__url__12__);
}
"
`;
